### ISSUE TITLE
Android support?

### ISSUE THREAD

**Issue by Rudxain** [undergroundwires/privacy.sexy#324]
### Problem description

3 OSes is nice, but including Android would be better

### Proposed solution

A checklist of settings to change manually

### Alternatives considered

Automated scripts would usually require `root` access (which is out-of-the-question for users who have fully-locked bootloaders). This is why I propose a manual checklist, as GUI Settings and ADB config cmds are more accessible.

### Additional information

Some scripts (such as shell history clear) can be executed on Termux without elevated privileges, and those are already included in the "Linux" section

**Comment by undergroundwires** [comment]
Hi, so we have two ways to go:

1. Provide scripts for Termux.
2. Provide scripts to run through ADB.

- Correct me if I'm wrong here but Termux is very limited on non-rooted devices right? It cannot even delete apps.
- With ADB, we can easily build a "remove bloatware" section for different brands. But is there more we can do there for non-rooted devices?

So I wonder which option should be more suitable, and what we can do on non-rooted devices with each option. Seems like we can do more with ADB.

**Comment by femdiya** [comment]
> Hi, so we have two ways to go:
> 
>     1. Provide scripts for Termux.
> 
>     2. Provide scripts to run through ADB.
> 
> 
>     * Correct me if I'm wrong here but Termux is very limited on non-rooted devices right? It cannot even delete apps.
> 
>     * With ADB, we can easily build a "remove bloatware" section for different brands. But is there more we can do there for non-rooted devices?
> 
> 
> So I wonder which option should be more suitable, and what we can do on non-rooted devices with each option. Seems like we can do more with ADB.

Android user here, 
Yes, Termux supports root but it's very limited on unrooted devices.
ADB is the best option here, but I suppose that user needs to execute the commands manually (?)

**Comment by undergroundwires** [comment]
On desktop version, we can provide a GUI/Wizard to go through ADB process. It may install ADB automatically or guide on how to install, select device, instruct what to do on mobile phone step by step through a GUI. We can run it when the user clicks on "Run" button. How does it sound?

On some scripts we can do root check using `adb`, if the device is rooted then execute otherwise show a warning, on scripts supporting non-root they can run directly.

Is there anything more that we can do through `adb` rather than "Remove bloatware" (remove packages) and "Privacy cleanup" (cleanup some folders)?

**Comment by Rudxain** [comment]
> Provide scripts to run through ADB.

[ADB can run on Termux](https://github.com/termux/termux-packages/tree/master/packages/android-tools). The setup process is clunkier because a ADB/TCP/IP/WiFi connection must be established. This must be done via USB (on older versions), but newer Android versions don't require USB for setup.

> Termux is very limited on non-rooted devices right?

Yes, but [there's a plugin to access the Android API](https://github.com/termux/termux-api), although it's much more limited than other apps such as Tasker, Macrodroid, or Llamalab's Automate.

> "remove bloatware" section for different brands.

[UAD](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation) aims to solve this. It doesn't have scripts, but users can select all packages from some category, then disable them all via ADB-over-USB.

> is there more we can do there for non-rooted devices?

Settings. I bet Xiaomi has tons of privacy-invasive settings. Samsung and Google both request usage stats, and have similar tracking settings. Since those settings are tied to the OS instead of an app, debloating alone can't help here.

Google Play Services has settings such as "Advertising ID", and that pack can't be disabled safely on some devices. Even when G.P.S. can be uninstalled/disabled, it provides APIs that most Play Store apps depend on. Discord doesn't even work at all when GPS is disabled (which is funny, because WhatsApp works ok, except for background messages), but I would argue that's the devs' fault for purposefully depending on it (the web app works fine without G.P.S.), and Discord+[uBO](https://github.com/gorhill/uBlock) is recommended anyways, so users shouldn't be using the native app

**Comment by Rudxain** [comment]
> Shizuku can be ran in termux

True. There's a button to "use in terminal apps" that provides a script that can be executed by Termux.

> just with a computer connected once

Shizuku setup is almost the same as ADB on Termux. There's no need for USB on newer Androids.

However, Shizuku makes pairing+connection less annoying:
- It shows an input notification that asks for the pairing PIN, which is better than multi-window.
- It auto-detects pairing and connection ports. I have no idea how it does it so fast, it cannot be port-scanning.

IIRC, Shizuku has a "keep-alive" feature that keeps Wireless-Debug enabled until the user disconnects (or wifi is turned off). Termux, OTOH, lets the OS disable WD after a period of inactivity.

> [!note]
> WD is the "developer setting", not `adb`, nor `shell` user

**Comment by undergroundwires** [comment]
I'm noob to Android and modding it. So I appreciate all discussions about it.

To be able to run `su` in Termux you need to root the phone right? Without rooting, it cannot configure the system,or delete files/apps, basically a sandboxed emulation of a unix shell, which is useless.

Looking at [`AndroidManifest.xml`](https://github.com/aosp-mirror/platform_frameworks_base/blob/master/packages/Shell/AndroidManifest.xml) we can see that a lot more permissions are granted to ADB.

Using ADB we can actually:

- Remove apps: `adb uninstall com.spotify.lite`.
- Disable background running: `adb shell cmd appops set <package_name> RUN_IN_BACKGROUND ignore`

ADB on Termux (which I did not know about as @Rudxain mentions) seems to remove the need for USB cable and PC connection and allow us to run `adb` CLI directly on Termux if I understand this correctly.

In that case we can  have some explanation on UI about how to get ADB running and then provide scripts utilizing ADB commands like (examples):
- Uninstall Xiaomi tracking: `adb uninstall com.huaqin.diaglogger`
- Uninstall Samsung tracking: `adb uninstall com.qualcomm.qti.qms.service.telemetry`

I'm not sure how to utilize Shizuku and if allows adb CLI usage as above.

I'm very interested in what other settings we can configure as you mentioned @Rudxain.

Does it make sense to create "Android" collection and adding these commands. To start with, we can add a dialog explaining how to get going with ADB using Termux or PC on the UI.

**Comment by Rudxain** [comment]
> you need to root the phone right?

Yes, but there's tools like [Magisk](https://github.com/topjohnwu/Magisk), which allow rooting without sacrificing too much of its benefits.

> Without rooting, it cannot configure the system

Well, the `settings` cmd (typically used via ADB) can be run on Termux, with [some caveats](https://github.com/termux/termux-packages/discussions/8292#discussioncomment-5102555).

> delete files/apps

If Termux (or any app) has "Access to all files" permission, its commands have it too. Of course, "All files" refers to user files. It cannot delete the data within other apps (such as cache).

Apps can request a "Request delete package" permission. But, as the name suggests, it requires manual user confirmation. Termux doesn't request this permission, but [LL-AM](https://llamalab.com/automate/) does. Termux can communicate with LL-AM in many ways ([Tasker plugin](https://github.com/termux/termux-tasker), user [file-system](https://llamalab.com/automate/doc/block/file_monitor.html), etc...), and we just need a "flow" for that.

> basically a sandboxed emulation of a unix shell

Sandboxed: yes. Emulated: no.

Apps can execute arbitrary programs (`execve` I guess?), as if they were Linux users (they are). They just need the proper permissions. Some programs may be executable, but have partial functionality.

[Apps targeting A10 cannot execute writable storage](https://github.com/termux/termux-app/issues/1072). Basically `W^X` applies to file-system as well as memory.

Android's shell is [MKSH](http://www.mirbsd.org/htman/i386/man1/mksh.htm), but Termux uses Bash by default.

> I'm not sure how to utilize Shizuku and if allows adb CLI usage as above.

The script I mentioned should support that. I don't remember how to use it. The Shizuku app documents it. IDK why the docs aren't available online.

Here's the [developer docs](https://github.com/RikkaApps/Shizuku-API), just-in-case I missed something.

> I'm very interested in what other settings we can configure

**AppOps** unlocks many possibilities! Such as disabling clipboard access, background location, etc...

Shizuku dev made [an app to manage them](https://play.google.com/store/apps/details?id=rikka.appops).

> Does it make sense to create "Android" collection

Anything Android-specific should be there. Cross-distro stuff is already handled by the Linux collection

**Comment by Rudxain** [comment]
> other settings

There are OEM-specific settings. [Samsung has had **a lot**](https://github.com/ionuttbara/melody_android) (~200) in the history of Galaxy devices.

Apps like LL-AM expose hidden features such as `secure` (there's also `global` and `system`) settings, unreachable app activities, services and broadcasts, etc... I recommend that sort of apps (such as Tasker and Macrodroid) to get a glimpse of the hidden stuff.

`WRITE_SECURE_SETTINGS` can be requested in the `AndroidManifest.xml`, but needs a one-time `adb shell pm grant`, and (for some reason) LL-AM cannot configure arbitrary `secure` settings (such as [display resolution](https://github.com/tytydraco/Purr), unless calling `adb shell wm size`)

**Comment by undergroundwires** [comment]
### ADB commands

🔍🧐 Please provide any docs/projects about what we can configure regarding privacy using `adb` on Android.

I think we decided that we'll provide a script collection based on `adb` commands and we have the consensus on this as way forward.

@Rudxain, I checked @ionuttbara's melody_android and it looks nice and provides some nice tweaks.

### Mobile app - Packaging

@plantindesk suggests in https://github.com/undergroundwires/privacy.sexy/issues/353#issuecomment-2088813411 that:

> We can use Native Script (Free) and Expo (Free-Limited in deployment) for android support 
> As all is dependent upon vue Nativescript is best

This is more about providing the Android package rather than the scripts. I checked Nativescript after your suggestion and it will require re-building the app, which would be too much.

I think publishing privacy.sexy  as PWA (progressive web app) would be the easiest option. [`vite-plugin-pwa`](https://vite-pwa-org.netlify.app/) seems to simplify this. It can then be pushed to Play Store etc. Google provides a CLI ([bubblewarp](https://github.com/GoogleChromeLabs/bubblewrap)) to turn PWA to Android APK that can be published.

A PWA cannot **run the scripts** as I understand, but it can give more offline experience to copy paste the scripts into Termux. This is sad as desktop version adds desktop-native functionality but mobile version will not be able to do that, but rebuilding a complete new app using Nativescript is too much time consuming and is not on my radar.

---

I read your posts @Rudxain and try to follow. Almost all of this stuff is new to me, so I may be missing some points or misunderstanding stuff. Feel free to guide us through this implementation.

**Comment by Rudxain** [comment]
> PWA cannot **run the scripts**

I'm confident in that being true. AFAIK, PWAs (`bubblewrap`ed or not) cannot access the Android API (unless indirect, such as Web `File` API), let alone the Linux ABI (`execve`).
 
BTW, converting the existing website into a PWA [should be pretty straight-forward](https://developer.mozilla.org/en-US/docs/Web/Manifest). So the offline experience can be achieved easily

**Comment by Rudxain** [comment]
I just remembered that [ADB can work via `WebUSB`](https://github.com/GoogleChromeLabs/wadb), so I would expect wireless-debug to also work via [WebRTC](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API). Since ADB is just an application-layer-protocol, it works over:
- USB
- Ethernet
- TCP/IP
- Bluetooth (maybe?)

And indeed, [it does seem like it should work via `WebRTC`!](https://source.android.com/docs/devices/cuttlefish/webrtc)

BTW, last time I checked, Firefox refuses to support `WebUSB` because Mozilla deems it unsafe/insecure. Chromium browsers support it, though

**Comment by mziehm** [comment]
Changing captive portal to privacy friendly alternative or deactivating does not require root
based on this info in German only: https://www.kuketz-blog.de/empfehlungsecke/#captive-portal:

**setting an alternative privacy friendly captive portal**:
```
adb shell 'settings put global captive_portal_http_url "http://captiveportal.kuketz.de"'
adb shell 'settings put global captive_portal_https_url "https://captiveportal.kuketz.de"'
adb shell 'settings put global captive_portal_fallback_url "http://captiveportal.kuketz.de"'
adb shell 'settings put global captive_portal_other_fallback_urls "http://captiveportal.kuketz.de"'
```

**deactivating  captive portal checks completely** (probably not recommended for most users):
```
adb shell 'settings put global captive_portal_detection_enabled 0'
adb shell 'settings put global captive_portal_server localhost'
adb shell 'settings put global captive_portal_mode 0'
```

**Comment by Rudxain** [comment]
> Changing captive portal to privacy friendly alternative

Aren't captive-portals supposed to be router-specific? How does that work?

**Comment by mziehm** [comment]
> > Changing captive portal to privacy friendly alternative
> Aren't captive-portals supposed to be router-specific? How does that work?

Sorry might not have formulated it precisely enough (made a small update above). I think, you are correct the captive portal is from the router. This is about the captive portal detection or check which is done by many clients incl. Windows, Android with the aim to detect if there is a captive portal instead of internet access. See also here:
https://en.wikipedia.org/wiki/Captive_portal#Detection

"Every time an Android device connects to a Wi-Fi or mobile network, the system performs a captive portal check. This is Android's way of ensuring that a device has not only received an IP address from the access point or Internet service provider, but can also actually reach destinations on the Internet. This is necessary, for example, to recognize portal pages of wifi hotspots (e.g. in hotels). 
To check whether a connection exists, Android sends a request to the address connectivitycheck.gstatic.com. If the request is successful, there is a connection to the Internet. With this request, the system transmits information about the IP address of the connection, the time of Internet access and the browser currently being used to Google each time. However, privacy-sensitive users in particular do not want to send a "ping" to Google every time they visit the Internet."
Translated with DeepL.com (free version) from https://www.kuketz-blog.de/empfehlungsecke/#captive-portal

**Comment by mziehm** [comment]
Privacy policy of the captiveportal.kuketz.de
"A connectivity check for Android smartphones is offered at captiveportal.kuketz.de. Anyone accessing the page receives the following response:

HTTP/1.1 204 No Content

Use of the service is voluntary. No data is collected in any form. After the request has been answered by the nginx web server, all information such as the IP address is immediately discarded and not stored in any log file."

Translated with DeepL.com (free version) from https://www.kuketz-blog.de/datenschutzhinweise/ paragraph "2.6 Android Captive Portal Check: captiveportal.kuketz.de"

There might/likely are other privacy friendly operators of such services. Never searched for it.

**Comment by femdiya** [comment]
Speaking of "privacy friendly captive portal", I think this website gets the job done and it's a great choice:
http://httpforever.com/
@mziehm @Rudxain

**Comment by brinerustle** [comment]
I'd argue that the only way to really get privacy on an android device is to degoogle it.

**Comment by Rudxain** [comment]
Should I open a new issue for this? I was considering the app/website could have a "user awareness" thing, because there are things that a user can easily avoid but are "unreachable" to ADB.

For example, any non-[`dangerous`](https://developer.android.com/guide/topics/permissions/overview#runtime) permission is granted at install-time, and I remember there are sensitive APIs a package can use **without explicitly requesting ANY permission!** To see what I mean, [install AIDA64](https://play.google.com/store/apps/details?id=com.finalwire.aida64), it can fetch **a lot of data** that can be trivially fingerprinted.

[I've elaborated on this in the UADNG guild](https://discord.com/channels/1168607797081022514/1285944507317354587/1346992366590099467). Should I quote some messages here or on a new issue?

[More info](https://consumerrights.wiki/w/Android_data_collection)

**Comment by HyperCriSiS** [comment]
This project could be used as resource regarding app debloating:
https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation

**Comment by Rudxain** [comment]
I've found [this app](https://github.com/lavafroth/droidrunco) which takes a "hybrid" approach in some sense. Instead of running completely in a browser, it runs natively on the host OS and opens a server whose GUI can be rendered by a browser. Unlike the pure web-app approach, this has minimal limitations! (no need for WebRTC or WebUSB, just syscalls). And because the GUI is provided by a browser, it should be accessible and multi-platform

**Comment by femdiya** [comment]
> This project could be used as resource regarding app debloating: https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation

https://github.com/samolego/Canta
This one has a good database too.

**Comment by Rudxain** [comment]
Canta depends on the UADNG DB 😅

**Comment by Rudxain** [comment]
Not sure if this is TCL specific, but on my TCL-505 the cmd
```sh
adb shell settings list system \| grep \^full_screen_
```
prints the package name/ID of pretty much **every package that was ever installed**. Mostly system apps, but there are also third-party apps I had **uninstalled months ago!**

And before anyone says "apps can't see that data because it needs ADB", I tested [this "flow"](https://llamalab.com/automate/community/flows/37254) (with 0 permissions) and it enumerated the same entries!

IMO, this is worse than [this exploit](https://peabee.substack.com/p/everyone-knows-what-apps-you-use)

I did a cleanup with this cmd
```sh
adb shell 'settings list system | grep \^full_screen_ | cut -d= -f1 | xargs -n1 settings delete system'
```
